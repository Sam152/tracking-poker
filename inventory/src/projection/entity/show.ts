import { StampedType } from "../../util/nominalType";

export type ShowId = StampedType<string, "ShowId">;
export type OperatorId = StampedType<string, "OperatorId">;
export type ShowDate = `${string}/${string}/${string}`;

export type Show = {
    id: ShowId;
    show_name: string;
    operator: OperatorId;
    date: ShowDate;
    duration: number;
};

export type ShowStorage = Show & {
    entity_type: "show";
    pk: `operator#${OperatorId}`;
    sk: `show#date#${ShowDate}#slug#${ShowId}#`;
    gsi1pk: `slug#${ShowId}`;
    gsi1sk: "show#";
};

export function toStorage(show: Show): ShowStorage {
    return {
        ...show,
        entity_type: "show",
        pk: `operator#${show.operator}`,
        sk: `show#date#${show.date}#slug#${show.id}#`,
        gsi1pk: `slug#${show.id}`,
        gsi1sk: "show#",
    };
}

// @todo, validate data from storage.
export function fromStorage(showStorage: Record<string, any>): Show {
    return {
        id: showStorage.id!,
        show_name: showStorage.show_name!,
        operator: showStorage.operator!,
        date: showStorage.date!,
        duration: showStorage.duration!,
    };
}
