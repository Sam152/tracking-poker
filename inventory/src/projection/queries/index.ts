import { QueryCommand } from "@aws-sdk/lib-dynamodb";

export type StructuredQuery<T> = {
    query: QueryCommand;
    parser: (items: Record<string, any>[]) => T;
};
