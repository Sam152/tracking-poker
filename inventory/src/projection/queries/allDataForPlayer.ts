import { QueryCommand } from "@aws-sdk/lib-dynamodb";
import { StructuredQuery } from "./index";
import * as Stat from "../entity/stat";
import { statGroupFromItems } from "../entity/stat";
import * as Player from "../entity/playerAppearance";
import * as PlayerAppearance from "../entity/playerAppearance";

type StatGroup = Partial<Record<Stat.StatType, Stat.Stat[]>>;

export function allDataForPlayer(player: Player.PlayerId): StructuredQuery<{
    appearances: PlayerAppearance.PlayerAppearance[];
    stats: StatGroup;
}> {
    return {
        query: new QueryCommand({
            TableName: "inventory",
            KeyConditionExpression: "pk = :pk",
            ExpressionAttributeValues: {
                ":pk": `player#${player}`,
            },
        }),
        parser: (items: Record<string, any>[]) => {
            return {
                appearances: items
                    .filter((item) => item.entity_type === "player_appearance")
                    .map((appearanceStorage) => PlayerAppearance.fromStorage(appearanceStorage)),
                stats: statGroupFromItems(items),
            };
        },
    };
}
