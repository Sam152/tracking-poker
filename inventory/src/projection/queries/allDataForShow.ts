import { QueryCommand } from "@aws-sdk/lib-dynamodb";
import { StructuredQuery } from "./index";
import * as Stat from "../entity/stat";
import { statGroupFromItems } from "../entity/stat";
import * as Show from "../entity/show";
import * as PlayerAppearance from "../entity/playerAppearance";

type StatGroup = Partial<Record<Stat.StatType, Stat.Stat[]>>;

export function allDataForShow(showId: Show.ShowId): StructuredQuery<{
    show?: Show.Show;
    players: PlayerAppearance.PlayerAppearance[];
    stats: StatGroup;
}> {
    return {
        query: new QueryCommand({
            TableName: "inventory",
            IndexName: "gsi1",
            KeyConditionExpression: "gsi1pk = :gsi1pk",
            ExpressionAttributeValues: {
                ":gsi1pk": `slug#${showId}`,
            },
        }),
        parser: (items: Record<string, any>[]) => {
            const show = items.find((item) => item.entity_type === "show");
            return {
                show: show ? Show.fromStorage(show) : undefined,
                players: items
                    .filter((item) => item.entity_type === "player_appearance")
                    .map((appearanceStorage) => PlayerAppearance.fromStorage(appearanceStorage)),
                stats: statGroupFromItems(items),
            };
        },
    };
}
