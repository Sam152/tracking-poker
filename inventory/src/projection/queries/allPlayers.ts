import { QueryCommand } from "@aws-sdk/lib-dynamodb";
import { StructuredQuery } from "./index";
import * as Player from "../entity/player";

export function allPlayers(): StructuredQuery<{
    players: Player.Player[];
}> {
    return {
        query: new QueryCommand({
            TableName: "inventory",
            KeyConditionExpression: "pk = :pk AND begins_with(sk, :skBeginsWith)",
            ExpressionAttributeValues: {
                ":pk": `all_operators`,
                ":skBeginsWith": `player#`,
            },
        }),
        parser: (items: Record<string, any>[]) => {
            return {
                players: items
                    .filter((item) => item.entity_type === "player")
                    .map((playerStorage) => Player.fromStorage(playerStorage)),
            };
        },
    };
}
