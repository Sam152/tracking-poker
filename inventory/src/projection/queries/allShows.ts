import { QueryCommand } from "@aws-sdk/lib-dynamodb";
import { StructuredQuery } from "./index";
import * as Show from "../entity/show";

export function allShows(): StructuredQuery<Show.Show[]> {
    return {
        query: new QueryCommand({
            // Scan index backwards, to show most recent shows first.
            ScanIndexForward: false,
            TableName: "inventory",
            KeyConditionExpression: "pk = :pk AND begins_with(sk, :skBeginsWith)",
            ExpressionAttributeValues: {
                ":pk": `operator#hcl`,
                ":skBeginsWith": `show#`,
            },
        }),
        parser: (items: Record<string, any>[]) => {
            return items.map((storageShow) => Show.fromStorage(storageShow));
        },
    };
}
