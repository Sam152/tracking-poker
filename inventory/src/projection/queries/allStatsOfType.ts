import { QueryCommand } from "@aws-sdk/lib-dynamodb";
import { StructuredQuery } from "./index";
import * as Stat from "../entity/stat";

export function allStatsOfType(type: Stat.StatType): StructuredQuery<Stat.Stat[]> {
    return {
        query: new QueryCommand({
            TableName: "inventory",
            IndexName: "gsi2",
            KeyConditionExpression: "gsi2pk = :gsi2pk",
            ExpressionAttributeValues: {
                ":gsi2pk": `stat_type#${type}`,
            },
        }),
        parser: (items: Record<string, any>[]) => {
            return items.map((statStorage) => Stat.fromStorage(statStorage));
        },
    };
}
