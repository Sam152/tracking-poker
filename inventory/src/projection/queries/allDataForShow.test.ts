import { stamp } from "../../util/nominalType";
import { allDataForShow } from "./allDataForShow";
import { ShowId } from "../entity/show";
import { allDataForShowFixture } from "../__fixtures__/queries/allDataForShowFixture";

describe("allDataForShow", () => {
    test("parser", () => {
        expect(allDataForShow(stamp<ShowId>("dVs7ORBHsF0")).parser(allDataForShowFixture)).toMatchSnapshot();
    });
});
