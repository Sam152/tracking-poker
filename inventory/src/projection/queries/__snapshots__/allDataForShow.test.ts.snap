// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`allDataForShow parser 1`] = `
{
  "players": [
    {
      "date": "2023/06/15",
      "player": "alex",
      "player_name": "ALEX",
      "show": "dVs7ORBHsF0",
      "show_name": "Sashimi, Kiran C2B, Israeli Ron & Mike X Play $10/20/40 - Commentary by Christian Soto",
    },
    {
      "date": "2023/06/15",
      "player": "c2b",
      "player_name": "C2B",
      "show": "dVs7ORBHsF0",
      "show_name": "Sashimi, Kiran C2B, Israeli Ron & Mike X Play $10/20/40 - Commentary by Christian Soto",
    },
    {
      "date": "2023/06/15",
      "player": "faizal",
      "player_name": "FAIZAL",
      "show": "dVs7ORBHsF0",
      "show_name": "Sashimi, Kiran C2B, Israeli Ron & Mike X Play $10/20/40 - Commentary by Christian Soto",
    },
    {
      "date": "2023/06/15",
      "player": "israeliron",
      "player_name": "ISRAELI RON",
      "show": "dVs7ORBHsF0",
      "show_name": "Sashimi, Kiran C2B, Israeli Ron & Mike X Play $10/20/40 - Commentary by Christian Soto",
    },
    {
      "date": "2023/06/15",
      "player": "johnny",
      "player_name": "JOHNNY",
      "show": "dVs7ORBHsF0",
      "show_name": "Sashimi, Kiran C2B, Israeli Ron & Mike X Play $10/20/40 - Commentary by Christian Soto",
    },
    {
      "date": "2023/06/15",
      "player": "mikex",
      "player_name": "MIKE X",
      "show": "dVs7ORBHsF0",
      "show_name": "Sashimi, Kiran C2B, Israeli Ron & Mike X Play $10/20/40 - Commentary by Christian Soto",
    },
    {
      "date": "2023/06/15",
      "player": "nick",
      "player_name": "NICK",
      "show": "dVs7ORBHsF0",
      "show_name": "Sashimi, Kiran C2B, Israeli Ron & Mike X Play $10/20/40 - Commentary by Christian Soto",
    },
    {
      "date": "2023/06/15",
      "player": "sashimi",
      "player_name": "SASHIMI",
      "show": "dVs7ORBHsF0",
      "show_name": "Sashimi, Kiran C2B, Israeli Ron & Mike X Play $10/20/40 - Commentary by Christian Soto",
    },
    {
      "date": "2023/06/15",
      "player": "tal",
      "player_name": "TAL",
      "show": "dVs7ORBHsF0",
      "show_name": "Sashimi, Kiran C2B, Israeli Ron & Mike X Play $10/20/40 - Commentary by Christian Soto",
    },
    {
      "date": "2023/06/15",
      "player": "wei",
      "player_name": "WEI",
      "show": "dVs7ORBHsF0",
      "show_name": "Sashimi, Kiran C2B, Israeli Ron & Mike X Play $10/20/40 - Commentary by Christian Soto",
    },
  ],
  "show": {
    "date": "2023/06/15",
    "duration": 20165,
    "id": "dVs7ORBHsF0",
    "operator": "hcl",
    "show_name": "Sashimi, Kiran C2B, Israeli Ron & Mike X Play $10/20/40 - Commentary by Christian Soto",
  },
  "stats": {
    "cc": [
      {
        "player": "alex",
        "player_name": "ALEX",
        "show": "dVs7ORBHsF0",
        "type": "cc",
        "value": 37360,
      },
      {
        "player": "c2b",
        "player_name": "C2B",
        "show": "dVs7ORBHsF0",
        "type": "cc",
        "value": 21390,
      },
      {
        "player": "faizal",
        "player_name": "FAIZAL",
        "show": "dVs7ORBHsF0",
        "type": "cc",
        "value": 17120,
      },
      {
        "player": "johnny",
        "player_name": "JOHNNY",
        "show": "dVs7ORBHsF0",
        "type": "cc",
        "value": 42480,
      },
      {
        "player": "mikex",
        "player_name": "MIKE X",
        "show": "dVs7ORBHsF0",
        "type": "cc",
        "value": 17160,
      },
      {
        "player": "nick",
        "player_name": "NICK",
        "show": "dVs7ORBHsF0",
        "type": "cc",
        "value": 24680,
      },
      {
        "player": "sashimi",
        "player_name": "SASHIMI",
        "show": "dVs7ORBHsF0",
        "type": "cc",
        "value": 34840,
      },
      {
        "player": "tal",
        "player_name": "TAL",
        "show": "dVs7ORBHsF0",
        "type": "cc",
        "value": 26410,
      },
      {
        "player": "wei",
        "player_name": "WEI",
        "show": "dVs7ORBHsF0",
        "type": "cc",
        "value": 8070,
      },
    ],
    "cw": [
      {
        "player": "alex",
        "player_name": "ALEX",
        "show": "dVs7ORBHsF0",
        "type": "cw",
        "value": 27530,
      },
      {
        "player": "c2b",
        "player_name": "C2B",
        "show": "dVs7ORBHsF0",
        "type": "cw",
        "value": 11760,
      },
      {
        "player": "faizal",
        "player_name": "FAIZAL",
        "show": "dVs7ORBHsF0",
        "type": "cw",
        "value": -5180,
      },
      {
        "player": "johnny",
        "player_name": "JOHNNY",
        "show": "dVs7ORBHsF0",
        "type": "cw",
        "value": 12480,
      },
      {
        "player": "mikex",
        "player_name": "MIKE X",
        "show": "dVs7ORBHsF0",
        "type": "cw",
        "value": -21840,
      },
      {
        "player": "nick",
        "player_name": "NICK",
        "show": "dVs7ORBHsF0",
        "type": "cw",
        "value": 5540,
      },
      {
        "player": "sashimi",
        "player_name": "SASHIMI",
        "show": "dVs7ORBHsF0",
        "type": "cw",
        "value": 5840,
      },
      {
        "player": "tal",
        "player_name": "TAL",
        "show": "dVs7ORBHsF0",
        "type": "cw",
        "value": -13560,
      },
      {
        "player": "wei",
        "player_name": "WEI",
        "show": "dVs7ORBHsF0",
        "type": "cw",
        "value": 2290,
      },
    ],
    "pfr": [
      {
        "player": "alex",
        "player_name": "ALEX",
        "show": "dVs7ORBHsF0",
        "type": "pfr",
        "value": 12,
      },
      {
        "player": "c2b",
        "player_name": "C2B",
        "show": "dVs7ORBHsF0",
        "type": "pfr",
        "value": 21,
      },
      {
        "player": "faizal",
        "player_name": "FAIZAL",
        "show": "dVs7ORBHsF0",
        "type": "pfr",
        "value": 14,
      },
      {
        "player": "israeliron",
        "player_name": "ISRAELI RON",
        "show": "dVs7ORBHsF0",
        "type": "pfr",
        "value": 15,
      },
      {
        "player": "johnny",
        "player_name": "JOHNNY",
        "show": "dVs7ORBHsF0",
        "type": "pfr",
        "value": 12,
      },
      {
        "player": "mikex",
        "player_name": "MIKE X",
        "show": "dVs7ORBHsF0",
        "type": "pfr",
        "value": 25,
      },
      {
        "player": "nick",
        "player_name": "NICK",
        "show": "dVs7ORBHsF0",
        "type": "pfr",
        "value": 11,
      },
      {
        "player": "sashimi",
        "player_name": "SASHIMI",
        "show": "dVs7ORBHsF0",
        "type": "pfr",
        "value": 20,
      },
      {
        "player": "tal",
        "player_name": "TAL",
        "show": "dVs7ORBHsF0",
        "type": "pfr",
        "value": 21,
      },
      {
        "player": "wei",
        "player_name": "WEI",
        "show": "dVs7ORBHsF0",
        "type": "pfr",
        "value": 18,
      },
    ],
    "vpip": [
      {
        "player": "alex",
        "player_name": "ALEX",
        "show": "dVs7ORBHsF0",
        "type": "vpip",
        "value": 35,
      },
      {
        "player": "c2b",
        "player_name": "C2B",
        "show": "dVs7ORBHsF0",
        "type": "vpip",
        "value": 31,
      },
      {
        "player": "faizal",
        "player_name": "FAIZAL",
        "show": "dVs7ORBHsF0",
        "type": "vpip",
        "value": 59,
      },
      {
        "player": "israeliron",
        "player_name": "ISRAELI RON",
        "show": "dVs7ORBHsF0",
        "type": "vpip",
        "value": 45,
      },
      {
        "player": "johnny",
        "player_name": "JOHNNY",
        "show": "dVs7ORBHsF0",
        "type": "vpip",
        "value": 43,
      },
      {
        "player": "mikex",
        "player_name": "MIKE X",
        "show": "dVs7ORBHsF0",
        "type": "vpip",
        "value": 44,
      },
      {
        "player": "nick",
        "player_name": "NICK",
        "show": "dVs7ORBHsF0",
        "type": "vpip",
        "value": 23,
      },
      {
        "player": "sashimi",
        "player_name": "SASHIMI",
        "show": "dVs7ORBHsF0",
        "type": "vpip",
        "value": 36,
      },
      {
        "player": "tal",
        "player_name": "TAL",
        "show": "dVs7ORBHsF0",
        "type": "vpip",
        "value": 35,
      },
      {
        "player": "wei",
        "player_name": "WEI",
        "show": "dVs7ORBHsF0",
        "type": "vpip",
        "value": 27,
      },
    ],
  },
}
`;
